### ISSUE TITLE
(CLOSE ME AT THE END) Add a way to filter attributes at a model-level to model yaml

### ISSUE THREAD

**Issue by owlas** [lightdash/lightdash#5442]
Moved to milestone: [🕵🏽‍♂️ Attribute-based access control](https://github.com/lightdash/lightdash/milestone/92)

<details>
<summary>Old issue details</summary>

I want to add a model-level configuration which filters the model to a specific set of values _**always**_.

### What: Roughly, what does this look like in the product?

- [ ] users can apply filters to Explores in their dbt projects. 
- [ ] The filters are applied on a model level (i.e. under `model` -->`meta` --> `filters`)
- [ ] the filter syntax is the same as what's used here: https://github.com/lightdash/lightdash/issues/3126
- [ ] these filters cannot be changed and cannot be removed in the UI - these are actually hidden from the user in the UI. They'd only be visible in the compiled SQL. 
- [ ] ^ this would add a filter to every query against this schema (e.g. below it would be `created_at_date` >= 2022-05-21`)

```yaml
models:
  name: users
  meta:
    filters:
      -  sql: ${created_at_date} >= '2022-05-21' 
```

Places where we must use this filter:
- When compiling metric queries, we need to add these filters
- When autocompleting filter values, we need to apply these filters

We must ensure we're wrapping the attributes when we pass them into query, especially if users can later change their attributes. To avoid sql injection. 

### Questions:

1. Should we use a similar filter syntax to how we use with metric filters? (but then we have much more limited filter functionality... and how would we eventually support multiple filter types?)

```yaml
# myschema.yml
models:
  - name: users
      meta:
          filters:
            - ${username}: "oliver%"
```

Alternatively we could allow raw sql for the filter (see @jw-aleksandr-zolotukhin 's suggestion below):

```yml
filters:
  - sql: "users.client_name = ${ user.attributes.company }"
```

2. Should we have this as a more specific filter/parameter? This would make it easier to implement more access filter-specific things in the UI (e.g. you only have access to dta `x` because you're in team `y`).
If we did this, we'd onlybe able to accept user attributes in these filters (not any field value from the table)

e.g.
```yaml
models:
  - name: users
      meta:
          access_filter:
              - field: products.brand_name
              - user_attribute: company
```

### Out of scope

- multiple filter types (eventually, we will have `type` in the filter configuration like described [here](https://github.com/lightdash/lightdash/issues/1735)). For now, we will only have `fixed` filter types.:
    - there are three types of filters:
      - `default` filters are applied to the table, but can be overridden in the UI (the same way as dashboard saved filters can be overriden, for example).
      - `required` filters can have their values changed, but cannot be removed from the filters list in the UI
      - `fixed` filters cannot be changed and cannot be removed - these are actually hidden from the user in the UI. They'd only be visible in the compiled SQL.

- showing filters in the UI - in future we may want to allow the user to customise whether these filters show in the UI
- making filters editable - in future we may want to always have a filter always present but the user can change the actual value in the UI
- using liquid syntax (we'll do this in the next ticket)
- Joins, we don't apply always_filter to a if it's in a join. Only when it is the base table.

</details>

**Comment by TuringLovesDeathMetal** [comment]
oooh cool! 

This overlaps a lot with this: #1735

**Comment by ZeRego** [comment]
Can users get around this when using table calculations?

**Comment by owlas** [comment]
@ZeRego maybe - do you have an example?

This would be added to the most low level filters (in the first CTE) before any calculations so I believe everything should be filtered out.


@TuringLovesDeathMetal I thought about having something more user attribute specific but I liked that with the current approach you wouldn't need an exact match to a user attribute, which in some cases might make a better UX like (bit of a stupid example but)

```
filters:
  - email: '%@${ user.attributes.email_domain }'
```

So each user could have an `email_domain` attribute but it could still be used to filter a full email field.

**Comment by ZeRego** [comment]
How could I use a boolean/number/array attribute?

**Comment by jw-aleksandr-zolotukhin** [comment]
Would be really handy to have a custom sql expression, not limiting it to some enum of filter options. It's more flexible.
Also could unlock usage of semi-structured attributes and their values (see question above wrt arrays).

**Comment by owlas** [comment]
Good suggestion, we could add the option to provide raw sql instead:

```yaml
always_filter:
  sql: "users.client_name = ${ user.attributes.company }"
```

What do you think @jw-aleksandr-zolotukhin ?

**Comment by owlas** [comment]
`sql_filter` is implemented in #6349 

Leaving this issue here as a parent issue for all other filter types

**Comment by TuringLovesDeathMetal** [comment]
docs are here: https://docs.lightdash.com/references/user-attributes

Note the security limitations here: https://docs.lightdash.com/references/user-attributes#current-limitations
